EKS Prep

yakloinsteak opened the pull request:
## Description

Changes needed to support Delayed Jobs on EKS.
- DJ metrics recorded and exposed
- Script to check if database is available and migrations have run.
- cronjob in EKS support

https://github.com/greenriver/argocd-manifests/pull/1032

## Type of change
- [x] New feature (adds functionality)
- [x] Code clean-up / housekeeping
- [x] Dependency update (new gems)

## Checklist before requesting review
- [x] I have performed a self-review of my code
- [x] I have run the code that is being changed under ideal conditions, and it doesn't fail
- [x] My code includes comments and/or descriptive variable names to help other engineers understand the intent (or not applicable)
- [x] My code follows the style guidelines of this project (rubocop)
- [x] I have updated the documentation (or not applicable)
- [x] If it's not obvious how to test this change, I have provided testing instructions in this PR or the related issue

yakloinsteak committed:
Changes for EKS

- delayed job metrics
- build branch tag to make iterations quicker
- webrick for delayed job metrics
- database ready script
- Delayed jobs don't need assets I would expect
- Hopefuly it's this easy to get rid of lots of noisy log messages we don't care about in development
- support for EKS cronjobs
- fix weird file descriptor error on EKS (setup_logging.rb change)
- Always init metrics on boot
- adjust which capacity cron jobs go to

yakloinsteak committed:
QA/fixes continue

yakloinsteak committed:
load service account name

yakloinsteak committed:
cleanup

eanders commented on app/jobs/base_job.rb:
```diff
@@ -8,36 +8,45 @@ class BaseJob < ApplicationJob
 
-      unlock_job!(job.id)
+    def before_handler(job)
+      self.start_time = Time.now
```
It's probably fine since you're using it all in the same file and just getting the time diff, but we tend to prefer `Time.current` over `Time.now` to incorporate the timezone.

eanders commented on app/jobs/base_job.rb:
```diff
@@ -52,39 +61,4 @@ def before(job)
-
-  private
-
-  def running_in_correct_path?
-    return true unless File.exist?('config/exception_notifier.yml')
-
-    expected_path == STARTING_PATH
-  end
-
-  def notify_on_restart(msg)
-    Rails.logger.info msg
-    return unless File.exist?('config/exception_notifier.yml')
-
-    setup_notifier('DelayedJobRestarter')
-    @notifier.ping(msg) if @send_notifications
-  end
-
-  def expected_path
-    Rails.cache.fetch('deploy-dir') do
-      # A default for the first deployment and local development
-      # This should be set on deployment.
-      Delayed::Worker::Deployment.deployed_to
-    end
-  end
-
-  def unlock_job!(job_id)
-    a_t = Delayed::Job.arel_table
-    job_object = Delayed::Job.where(a_t[:handler].matches("%job_id: #{job_id}%").or(a_t[:id].eq(job_id))).first
-    return unless job_object
-
-    msg = "Restarting stale delayed job: #{job_object.locked_by}"
-    notify_on_restart(msg)
-
-    job_object.update(locked_by: nil, locked_at: nil)
```
Is this all code related to the bare-metal deployments?  It looks like we moved this logic to the `WorkerStatus` class for use with ECS, just want to confirm.

yakloinsteak replied:
I believe this was all related to bare-metal deployments. Are you agreeing that it's dead code? That's what I was thinking it was at least.

eanders replied:
Yes, I was agreeing with that assessment, just wanted to make sure that's why you were removing it.

eanders commented on config/deploy/docker/lib/cronjob.rb:
```diff
@@ -0,0 +1,152 @@
+
+  MAX_NAME_LENGTH = 52
+
+  def initialize(description:, schedule_expression:, command:, capacity_type: 'spot')
```
I haven't found it yet, so just noting, `capacity_provider` should probably not be spot if the task is not interruptible.

yakloinsteak replied:
I'm just making the default the cheaper option. The logic in `config/schedule.rb` is explicit about all of the entries I think. interruptible jobs are put on spot capacity. non-interruptible on on-demand.

eanders reviewed (approved):
@yakloinsteak I think this looks great.  Thank you.  Can we target `release-134` so it doesn't go out to production immediately?

yakloinsteak committed:
Fix Time call

yakloinsteak committed:
Merge remote-tracking branch 'origin/release-134' into tb/dj-metrics

yakloinsteak committed:
Debugging encoding problem on EKS

yakloinsteak committed:
metrics files are binary, so maybe this will help

yakloinsteak committed:
There are a lot of unnessesary AWS api calls to RDS

They slow down booting and aren't usually needed. This is an attempt to
improve the situation.

yakloinsteak committed:
Future memoize to reduce bootup time (TESTING)

yakloinsteak committed:
rollback debug

yakloinsteak committed:
Merge branch 'release-134' into tb/dj-metrics

yakloinsteak commented:
May I merge this today, @eanders ? You might want to peek at https://github.com/greenriver/hmis-warehouse/blob/tb/dj-metrics/lib/rds_sql_server/sql_server_base.rb where I've memoized some things to reduce AWS API hits for all of the subclasses of that class. I believe it will help with boot time. On that topic, I was wondering if we should combine some of the rake tasks in the deploy_task.sh script. I think it would speed things up to _not_ load the rails stack over and over.

eanders reviewed (approved):
Looks good to me, let's give it a try.
